### DISCUSSION TITLE
GraphQL API cannot find references nested outside "refs/heads" or "refs/tags" while REST API can

### DISCUSSION THREAD

**Discussion by jsbronder** [community/community#83980] [Apps, API and Webhooks]
### Select Topic Area

Bug

### Body

The GraphQL queries for references appear to be unable to find any references not nested under "refs/heads" or "refs/tags".  While probably not common, refs/<anything> is perfectly valid.  I assume that this is unintentional as when I use the REST API (https://docs.github.com/en/rest/git/refs?apiVersion=2022-11-28#list-matching-references) I do get the references I'm searching for.

Code to reproduce:
```bash
#!/bin/bash

owner="YOUR_NAME_HERE"
repo="YOUR_REPO_HERE"

echo "Pushing some example refs to origin"
git push origin origin/master:refs/example/1234/ref1 origin/master:refs/example/1234/ref2

echo;echo
echo "Searching with graphql"
q=$(cat <<-EOF
query findRefId {
  repository(owner: "${owner}", name: "${repo}") {
    refs(first: 100,  refPrefix:"refs/example/") {
      totalCount
      nodes {
        id name
      }
    }
  }
}
EOF
)

gh api graphql -f query="${q}" | jq

echo;echo
echo "Searching with REST"
gh api repos/${owner}/${repo}/git/matching-refs/example | jq
```

Actual output:
```
Pushing some example refs to origin
Everything up-to-date


Searching with graphql
{
  "data": {
    "repository": {
      "refs": {
        "totalCount": 0,
        "nodes": []
      }
    }
  }
}


Searching with REST
[
  {
    "ref": "refs/example/1234/ref1",
    ...
  },
  {
    "ref": "refs/example/1234/ref2",
    ...
  }
]
```

Expected output:
```
Pushing some example refs to origin
Everything up-to-date


Searching with graphql
{
  "data": {
    "repository": {
      "refs": {
        "totalCount": 2,
        "nodes": [
          {
            "name": "1234/ref1",
            ...
          },
          {
            "name": "1234/ref2",
            ...
          }
        ]
      }
    }
  }
}


Searching with REST
[
  {
    "ref": "refs/example/1234/ref1",
    ...
  },
  {
    "ref": "refs/example/1234/ref2",
    ...
  }
]
```

### Update
This appears to be a problem throughout the GraphQL API.  Even if I use the `node_id` of a reference I looked up with the REST API, I still cannot use it in GraphQL.  For instance, I cannot delete a reference.
```
$ gh api repos/my_user/my_repo/git/refs/example/1234/ref1 | jq .node_id
"REF_kwDOKwx6kLZyZWZzL2V4YW1wbGUvMTIzNC9yZWYx"
$ gh api graphql -f query="mutation { deleteRef(input: {refId: \"REF_kwDOKwx6kLZyZWZzL2V4YW1wbGUvMTIzNC9yZWYx\"}) { clientMutationId } }"
{
  "data": {
    "deleteRef": null
  },
  "errors": [
    {
      "type": "NOT_FOUND",
      "path": [
        "deleteRef"
      ],
      "locations": [
        {
          "line": 1,
          "column": 12
        }
      ],
      "message": "Could not resolve to a node with the global id of 'REF_kwDOKwx6kLZyZWZzL2V4YW1wbGUvMTIzNC9yZWYx'."
    }
  ]
}
```

**Comment by AboFasil** [comment]
>[[]()]()

**Comment by jsbronder** [comment]
I'm not sure how to update it, but the label here should probably be 'bug'.

**Comment by jsbronder** [comment]
Working as expected per GitHub support.

> GitHub, Jan 22, 2024, 12:36 PM UTC
> 
> Hi Justin,
> 
> Thanks for reaching out and sorry for the delay in getting back to you on this.
> 
> Yeah, the GraphQL API will currently return only refs from the `heads`, `guest`, `tags`, `notes` namespaces. We have an
> internal issue with engineering tracking similar reports as feedback and I have added yours as a `+1`.
> 
> I can't promise if or when this request will be implemented but I'd recommend you can keep an eye on our GitHub Blog
> and the Changelog  where we make announcements on any change/update.
> 
> Please let me know if you have any questions.
> 
> Regards,
> 
> Oluwaseun
> GitHub Support
